File: apps/api/app/schemas/job.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.milestone import MilestoneCreate, MilestoneRead


class JobBase(BaseModel):
	title: str
	description: str
	total_amount: Decimal = Field(alias="totalAmount")


class JobCreate(JobBase):
	model_config = ConfigDict(populate_by_name=True)

	freelancer_id: str | None = Field(default=None, alias="freelancerId")
	freelancer_stellar_address: str | None = Field(default=None, alias="freelancerStellarAddress")
	milestones: list[MilestoneCreate] = Field(default_factory=list)

	@model_validator(mode="after")
	def validate_milestone_amounts(self) -> "JobCreate":
		if self.milestones:
			total = sum((milestone.amount for milestone in self.milestones), Decimal("0"))
			if total != self.total_amount:
				raise ValueError("Milestone amounts must equal total_amount")
		return self


class JobUpdate(BaseModel):
	title: str | None = None
	description: str | None = None
	freelancer_id: str | None = Field(default=None, alias="freelancerId")


class JobRead(BaseModel):
	model_config = ConfigDict(from_attributes=True, populate_by_name=True)

	id: str
	client_id: str = Field(alias="clientId")
	freelancer_id: str | None = Field(default=None, alias="freelancerId")
	title: str
	description: str
	total_amount: Decimal = Field(alias="totalAmount")
	status: str
	escrow_id: str | None = Field(default=None, alias="escrowId")
	contract_address: str | None = Field(default=None, alias="contractAddress")
	milestones: list[MilestoneRead] = Field(default_factory=list)
	created_at: datetime = Field(alias="createdAt")
	updated_at: datetime = Field(alias="updatedAt")

File: apps/web/types/job.ts
/**
 * Job type — aligned with context.md and the FastAPI Pydantic schema.
 *
 * IMPORTANT: totalAmount and milestone amounts are strings, not numbers.
 * Stellar uses 7 decimal places (NUMERIC 18,7). Never use JS number for USDC.
 */

import type { Milestone } from "./milestone";

export type JobStatus =
  | "DRAFT"
  | "FUNDED"
  | "IN_PROGRESS"
  | "MILESTONE_SUBMITTED"
  | "DISPUTED"
  | "COMPLETED"
  | "RELEASED";

export interface Job {
  id: string;
  clientId: string;
  freelancerId: string | null;
  title: string;
  description: string;
  /** USDC amount as a string — never parse to float */
  totalAmount: string;
  status: JobStatus;
  /** Returned by Trustless Work on escrow init */
  escrowId: string | null;
  /** Soroban contract address on Stellar */
  contractAddress: string | null;
  milestones: Milestone[];
  createdAt: string;
  updatedAt: string;
}

/** Shape for creating a new job */
export interface JobCreate {
  title: string;
  description: string;
  totalAmount: string;
  freelancerEmail?: string;
  freelancerStellarAddress?: string;
  milestones: {
    title: string;
    description: string;
    amount: string;
  }[];
}
